### PR TITLE
planner: add new explain format for plan cache

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Issue Number: ref #36598

Problem Summary:

### What is changed and how it works?
When explain use `format = 'plan_cache'` the query can use non-prepared plan cache(if enabled), and show cache miss reason by warnings. Other explain format can't use cache and will not be affected by plan cache.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -154,6 +154,7 @@ type StatementContext struct {
 	InSelectStmt                  bool
 	InLoadDataStmt                bool
 	InExplainStmt                 bool
+	ExplainFormat                 string
```
Comment (qw4990) [inline sessionctx/stmtctx/stmtctx.go:157]: How about using `InExplainPlanCacheFormat bool`?

File: executor/executor.go [hunk 2]
```diff
@@ -2104,12 +2104,14 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
...
 	if explainStmt, ok := s.(*ast.ExplainStmt); ok {
 		sc.InExplainStmt = true
+		sc.ExplainFormat = explainStmt.Format
 		sc.IgnoreExplainIDSuffix = strings.ToLower(explainStmt.Format) == types.ExplainFormatBrief
 		sc.InVerboseExplain = strings.ToLower(explainStmt.Format) == types.ExplainFormatVerbose
 		s = explainStmt.Stmt
 	}
 	if explainForStmt, ok := s.(*ast.ExplainForStmt); ok {
 		sc.InExplainStmt = true
+		sc.ExplainFormat = explainForStmt.Format
```
Comment (qw4990) [inline executor/executor.go:2114]: `Explain For Conn` statement is special, we don't need to let it support `plan-cache` format, so this line can be just removed.

File: executor/executor.go [hunk 3]
```diff
@@ -2104,12 +2104,14 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 	sc.OriginalSQL = s.Text()
 	if explainStmt, ok := s.(*ast.ExplainStmt); ok {
 		sc.InExplainStmt = true
+		sc.ExplainFormat = explainStmt.Format
```
Comment (qw4990) [inline executor/executor.go:2107]: And we should set this field to an empty string "" if it's not a `ExplainStmt`.

File: executor/explain_test.go [hunk 4]
```diff
@@ -610,3 +611,90 @@ func TestExplainJSON(t *testing.T) {
...
+
+	tk.MustExec("explain analyze format = 'plan_cache' select * from t limit 1")
+	tk.MustQuery("show warnings").Check(testkit.Rows("Warning 1105 skip non-prep plan cache: queries that have hints, aggregation, window-function, order-by, limit and lock are not supported"))
+	tk.MustExec("explain analyze format = 'plan_cache' select * from t limit 1")
+	tk.MustQuery("select @@last_plan_from_cache").Check(testkit.Rows("0"))
+
+	connID := tk.MustQuery("select connection_id()").Rows()[0][0]
+	tk.MustExec("select * from t limit 1")
+	tk.MustExec(fmt.Sprintf("explain format = 'plan_cache' for connection %v", connID))
+	tk.MustQuery("show warnings").Check(testkit.Rows("Warning 1105 skip non-prep plan cache: not a select statement"))
```
Comment (qw4990) [inline executor/explain_test.go:675]: Don't need to support `explain for conn`.

File: executor/explain_test.go [hunk 5]
```diff
@@ -610,3 +611,90 @@ func TestExplainJSON(t *testing.T) {
...
+		types.ExplainFormatVerbose,
+		types.ExplainFormatTraditional,
+		types.ExplainFormatBinary,
+		types.ExplainFormatTiDBJSON,
+		types.ExplainFormatCostTrace,
+		types.ExplainFormatPlanCache,
+	}
+
+	for _, format := range explainFormats {
+		tk.MustExec(fmt.Sprintf("explain analyze format = '%v' select * from t", format))
```
[the remark this answers is not in the document]
Comment (fzzf678) [reply executor/explain_test.go:642]: All fixed, PTAL again

**Review by qw4990 — APPROVED** [review]
LGTM

**Review by Reminiscent — APPROVED** [review]

**Pull request merged**